src/core/layout/qgslayoutitemlegend.cpp

```diff
@@ -880,22 +890,79 @@ QgsLegendModel::QgsLegendModel( QgsLayerTree *rootNode, QObject *parent )
+  if ( nodeLayer && ( role == Qt::DisplayRole || role == Qt::EditRole ) )
   {
-    if ( QgsLayerTree::isLayer( node ) && ( role == Qt::DisplayRole || role == Qt::EditRole ) && !node->customProperty( QStringLiteral( "legend/title-label" ) ).isNull() )
+    QString name;
+    QgsSymbolLegendNode *symnode = nullptr;
+    QgsVectorLayer *vlayer = qobject_cast<QgsVectorLayer *>( nodeLayer->layer() );
+    if ( ltmln )
+    {
+      symnode = qobject_cast<QgsSymbolLegendNode *>( ltmln );
+      name = symnode->symbolLabel();
+    }
+    else
     {
-      QgsLayerTreeLayer *nodeLayer = QgsLayerTree::toLayer( node );
-      QString name = node->customProperty( QStringLiteral( "legend/title-label" ) ).toString();
-      if ( nodeLayer->customProperty( QStringLiteral( "showFeatureCount" ), 0 ).toInt() && role == Qt::DisplayRole )
+      //finding the first label that is stored
+      name = nodeLayer->customProperty( QStringLiteral( "legend/title-label" ) ).toString();
+      if ( name.isEmpty() )
+        name = nodeLayer->name();
+      if ( name.isEmpty() )
+        name = node->customProperty( QStringLiteral( "legend/title-label" ) ).toString();
+      if ( name.isEmpty() )
+        name = node->name();
+      if ( nodeLayer->customProperty( QStringLiteral( "showFeatureCount" ), 0 ).toInt() )
       {
-        QgsVectorLayer *vlayer = qobject_cast<QgsVectorLayer *>( nodeLayer->layer() );
         if ( vlayer && vlayer->featureCount() >= 0 )
+        {
           name += QStringLiteral( " [%1]" ).arg( vlayer->featureCount() );
+          Q_UNUSED( ltmln );
+          Q_UNUSED( symnode );
+          return name;
+        }
       }
-      return name;
     }
-  }
 
+    bool evaluate = vlayer ? !nodeLayer->labelExpression().isEmpty() : false;
+
+    if ( evaluate || name.contains( "[%" ) )
+    {
+      QgsExpressionContext expressionContext;
+      if ( vlayer )
+        connect( vlayer, &QgsVectorLayer::symbolFeatureCountMapChanged, this, &QgsLegendModel::forceRefresh );
+
+      if ( mLayoutLegend )
+        expressionContext = mLayoutLegend->createExpressionContext();
+      else if ( vlayer )
+        expressionContext = vlayer->createExpressionContext();
+      else
+        expressionContext = QgsExpressionContext();
+
+      if ( symnode )
+        name = symnode->evaluateLabel( expressionContext ); // removed name input; existing symbol/model tree have distinct names
+      else
+      {
+        const QList<QgsLayerTreeModelLegendNode *> legendnodes = layerLegendNodes( nodeLayer, false );
+        if ( legendnodes.count() > 1 ) // evaluate all existing legend nodes but leave the name for the legend evaluator
+        {
+          for ( QgsLayerTreeModelLegendNode *treenode : legendnodes )
+          {
+            if ( QgsSymbolLegendNode *symnode = qobject_cast<QgsSymbolLegendNode *>( treenode ) )
+            {
+              symnode->evaluateLabel( expressionContext );
+              expressionContext.clearCachedValues();
```

Comment:
Since it's passed const and internally a copy is taken, I don't think it's required to call this.

```suggestion
```

Reply:
done